File: src/main/java/at/rueckgr/spotlight/service/SpotlightDownloader.java
package at.rueckgr.spotlight.service;

import lombok.extern.log4j.Log4j2;

@Log4j2
public class SpotlightDownloader {
    private final DownloadDirectoryService downloadDirectoryService;
    private final ImageDataDownloaderService imageDataDownloaderService;
    private final MetadataService metadataService;
    private final MetricsService metricsService;

    public static final int ITERATIONS = 100;

    public SpotlightDownloader() {
        // using Dependency Injection, life would be easier
        final LocalesService localesService = new LocalesService();

        metricsService = new MetricsService();
        metadataService = new MetadataService(localesService);
        downloadDirectoryService = new DownloadDirectoryService(metadataService, metricsService);
        metadataService.setDownloadDirectoryService(downloadDirectoryService);
        imageDataDownloaderService = new ImageDataDownloaderService(localesService, metricsService);
    }

    public void run() {
        downloadDirectoryService.init();

        final Worker worker = new Worker(imageDataDownloaderService, downloadDirectoryService);
        for(int i = 0; i < ITERATIONS; i++) {
            log.info("Iteration {}/{} done", i, ITERATIONS);
            worker.run();
        }
        log.info("All iterations done");

        metadataService.saveMetadata();
        metricsService.logMetrics();
    }
}

File: src/main/kotlin/at/rueckgr/spotlight/service/SpotlightDownloader.kt
package at.rueckgr.spotlight.service

import org.apache.logging.log4j.LogManager
import kotlin.system.exitProcess

fun main() {
    val log = LogManager.getLogger("Main")

    try {
        val iterations = 100

        val localesService = LocalesService()
        val metricsService = MetricsService()
        val metadataService = MetadataService(localesService)
        val downloadDirectoryService = DownloadDirectoryService(metadataService, metricsService)
        metadataService.downloadDirectoryService = downloadDirectoryService
        val imageDataDownloaderService = ImageDataDownloaderService(localesService, metricsService)

        downloadDirectoryService.init()
        val worker = Worker(imageDataDownloaderService, downloadDirectoryService)
        log.info("Starting download")
        for (i in 1..iterations) {
            worker.run()
            log.info("Iteration {}/{} done", i, iterations)
        }

        metadataService.saveMetadata()
        metricsService.logMetrics()

        if (metricsService.getMetric(MetricsService.Metric.DOWNLOAD_IMAGE) == 0) {
            log.error("No image was downloaded")
            exitProcess(1)
        }
    }
    catch (e: Throwable) {
        log.error("Exception occurred", e)
        exitProcess(1)
    }
}
